File: app/src/main/java/com/example/phone_stats_collector/UpdateChecker.java
package com.example.phone_stats_collector;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

public class UpdateChecker {
    private static final String TAG = "UpdateChecker";
    private static final String GITHUB_API =
            "https://api.github.com/repos/choyeun/phone-stats-collector/releases/latest";

    public static class UpdateInfo {
        public final boolean hasUpdate;
        public final String latestVersion;
        public final String downloadUrl;
        public final long apkSize;
        public final String releaseBody;

        public UpdateInfo(boolean hasUpdate, String latestVersion,
                          String downloadUrl, long apkSize, String releaseBody) {
            this.hasUpdate = hasUpdate;
            this.latestVersion = latestVersion;
            this.downloadUrl = downloadUrl;
            this.apkSize = apkSize;
            this.releaseBody = releaseBody;
        }

        public static UpdateInfo none() {
            return new UpdateInfo(false, null, null, 0, null);
        }
    }

    public static UpdateInfo check(String currentVersion) {
        try {
            HttpURLConnection conn = (HttpURLConnection) new URL(GITHUB_API).openConnection();
            conn.setRequestMethod("GET");
            conn.setRequestProperty("Accept", "application/vnd.github.v3+json");
            conn.setConnectTimeout(5000);
            conn.setReadTimeout(5000);

            int code = conn.getResponseCode();
            if (code != 200) {
                Log.w(TAG, "GitHub API 응답 " + code);
                return UpdateInfo.none();
            }

            BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream()));
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = br.readLine()) != null) sb.append(line);
            br.close();
            conn.disconnect();

            JSONObject json = new JSONObject(sb.toString());
            String tagName = json.optString("tag_name", "");
            String body = json.optString("body", "");
            String latestVer = tagName.startsWith("v") ? tagName.substring(1) : tagName;

            String downloadUrl = null;
            long apkSize = 0;
            JSONArray assets = json.optJSONArray("assets");
            if (assets != null) {
                for (int i = 0; i < assets.length(); i++) {
                    JSONObject asset = assets.getJSONObject(i);
                    String name = asset.optString("name", "");
                    if (name.endsWith(".apk")) {
                        downloadUrl = asset.optString("browser_download_url", null);
                        apkSize = asset.optLong("size", 0);
                        break;
                    }
                }
            }
            if (downloadUrl == null) {
                downloadUrl = "https://github.com/choyeun/phone-stats-collector/releases/download/"
                        + tagName + "/phone-stats-collector-" + tagName + ".apk";
            }

            boolean hasUpdate = isNewer(currentVersion, latestVer);
            Log.d(TAG, String.format("현재=%s 최신=%s 업데이트%s",
                    currentVersion, latestVer, hasUpdate ? "있음 ✅" : "없음"));

            return new UpdateInfo(hasUpdate, latestVer, downloadUrl, apkSize, body);

        } catch (Exception e) {
            Log.e(TAG, "버전 체크 실패: " + e.getMessage());
            return UpdateInfo.none();
        }
    }

    public static boolean isNewer(String current, String latest) {
        int[] cur = parseVersion(current);
        int[] lat = parseVersion(latest);
        if (cur.length == 0 || lat.length == 0) return false;

        int len = Math.max(cur.length, lat.length);
        for (int i = 0; i < len; i++) {
            int c = i < cur.length ? cur[i] : 0;
            int l = i < lat.length ? lat[i] : 0;
            if (l > c) return true;
            if (l < c) return false;
        }
        return false;
    }

    public static int[] parseVersion(String v) {
        if (v == null || v.isEmpty()) return new int[0];
        try {
            String[] parts = v.split("\\.");
            int[] result = new int[parts.length];
            for (int i = 0; i < parts.length; i++) {
                result[i] = Integer.parseInt(parts[i]);
            }
            return result;
        } catch (NumberFormatException e) {
            return new int[0];
        }
    }
}

File: app/src/main/java/com/example/phone_stats_collector/UpdateChecker.kt
package com.example.phone_stats_collector

import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

object UpdateChecker {
    private const val TAG = "UpdateChecker"
    private const val GITHUB_API =
        "https://api.github.com/repos/choyeun/phone-stats-collector/releases/latest"

    data class UpdateInfo(
        val hasUpdate: Boolean,
        val latestVersion: String?,
        val downloadUrl: String?,
        val apkSize: Long,
        val releaseBody: String?
    ) {
        companion object {
            fun none() = UpdateInfo(false, null, null, 0, null)
        }
    }

    fun check(currentVersion: String): UpdateInfo {
        try {
            val conn = URL(GITHUB_API).openConnection() as HttpURLConnection
            conn.requestMethod = "GET"
            conn.setRequestProperty("Accept", "application/vnd.github.v3+json")
            conn.connectTimeout = 5000
            conn.readTimeout = 5000

            val code = conn.responseCode
            if (code != 200) {
                Log.w(TAG, "GitHub API 응답 $code")
                return UpdateInfo.none()
            }

            val reader = BufferedReader(InputStreamReader(conn.inputStream))
            val sb = StringBuilder()
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                sb.append(line)
            }
            reader.close()
            conn.disconnect()

            val json = JSONObject(sb.toString())
            val tagName = json.optString("tag_name", "")
            val body = json.optString("body", "")
            val latestVer = if (tagName.startsWith("v")) tagName.substring(1) else tagName

            var downloadUrl: String? = null
            var apkSize: Long = 0
            val assets = json.optJSONArray("assets")
            if (assets != null) {
                for (i in 0 until assets.length()) {
                    val asset = assets.getJSONObject(i)
                    val name = asset.optString("name", "")
                    if (name.endsWith(".apk")) {
                        downloadUrl = asset.optString("browser_download_url", null)
                        apkSize = asset.optLong("size", 0)
                        break
                    }
                }
            }
            if (downloadUrl == null) {
                downloadUrl = "https://github.com/choyeun/phone-stats-collector/releases/download/$tagName/phone-stats-collector-$tagName.apk"
            }

            val hasUpdate = isNewer(currentVersion, latestVer)
            Log.d(TAG, "현재=$currentVersion 최신=$latestVer 업데이트${if (hasUpdate) "있음 ✅" else "없음"}")

            return UpdateInfo(hasUpdate, latestVer, downloadUrl, apkSize, body)

        } catch (e: Exception) {
            Log.e(TAG, "버전 체크 실패: ${e.message}")
            return UpdateInfo.none()
        }
    }

    fun isNewer(current: String, latest: String): Boolean {
        val cur = parseVersion(current)
        val lat = parseVersion(latest)
        if (cur.isEmpty() || lat.isEmpty()) return false

        val len = maxOf(cur.size, lat.size)
        for (i in 0 until len) {
            val c = if (i < cur.size) cur[i] else 0L
            val l = if (i < lat.size) lat[i] else 0L
            if (l > c) return true
            if (l < c) return false
        }
        return false
    }

    fun parseVersion(v: String?): LongArray {
        if (v.isNullOrEmpty()) return LongArray(0)
        return try {
            val parts = v.split("\\.".toRegex())
            LongArray(parts.size) { i ->
                parts[i].replace("[^0-9]".toRegex(), "").toLong()
            }
        } catch (_: NumberFormatException) {
            LongArray(0)
        }
    }
}